File: 557-reverse-words-in-a-string-iii/reverse-words-in-a-string-iii.py
class Solution:
    def reverseWords(self, s: str) -> str:
        words = s.split()
        for i in range(len(words)):
            word = list(words[i])
            L, R = 0, len(word) - 1
            while L < R:
                word[L], word[R] = word[R], word[L]
                L += 1
                R -= 1
            words[i] = "".join(word)
        return " ".join(words)

File: 557-reverse-words-in-a-string-iii/reverse-words-in-a-string-iii.cpp
class Solution {
public:
    string reverseWords(string s) {
        int lastSpaceIdx = -1;
        string res = "";
        int n = s.size();
        for (int i = 0; i < n; ++i) {
            if (i == n - 1 || s[i] == ' ') {
                int reverseIdx = (i == n - 1) ? i : i - 1;
                for (; reverseIdx > lastSpaceIdx; --reverseIdx) {
                    res += s[reverseIdx];
                }
                lastSpaceIdx = i;

                if (i != n - 1) {
                    res += ' ';
                }
            }
        }
        return res;
    }
};